### PR TITLE
docs(lua_ls): change suggested setup to work on both nightly and 0.9.5

### PR DESCRIPTION
**Pull request by abeldekat**
https://github.com/neovim/nvim-lspconfig/assets/58370433/49425c0c-9b6a-461c-a434-2341b337ef6f

In `neovim` nightly, the current suggested config fails. I added a video demo, using this repro:
```lua
--- CHANGE THESE
local pattern = "lua"
local cmd = { "lua-language-server" }
-- Add files/folders here that indicate the root of a project
local root_markers = { ".git", ".editorconfig" }
-- Change to table with settings if required
local settings = {
  Lua = {
    runtime = {
      -- Tell the language server which version of Lua you're using
      -- (most likely LuaJIT in the case of Neovim)
      version = "LuaJIT",
    },
    -- Make the server aware of Neovim runtime files
    workspace = {
      checkThirdParty = false,
      library = {
        vim.env.VIMRUNTIME,
        -- Depending on the usage, you might want to add additional paths here.
        -- E.g.: For using `vim.*` functions, add vim.env.VIMRUNTIME/lua.
        -- "${3rd}/luv/library"
        -- "${3rd}/busted/library",
      },
      -- or pull in all of 'runtimepath'. NOTE: this is a lot slower
      -- library = vim.api.nvim_get_runtime_file("", true)
    },
  },
}
local function on_init(client) -- recommended in nvim-lspconfig
  local path = client.workspace_folders[1].name
  if not vim.loop.fs_stat(path .. "/.luarc.json") and not vim.loop.fs_stat(path .. "/.luarc.jsonc") then
    client.config.settings = vim.tbl_deep_extend("force", client.config.settings, settings)
  end
  return true
end

vim.api.nvim_create_autocmd("FileType", {
  pattern = pattern,
  callback = function(args)
    local match = vim.fs.find(root_markers, { path = args.file, upward = true })[1]
    local root_dir = match and vim.fn.fnamemodify(match, ":p:h") or nil
    vim.lsp.start({
      name = "bugged-ls",
      cmd = cmd,
      root_dir = root_dir,
      on_init = on_init, -- either use on_init
      -- settings = settings, -- or the settings directly
    })
  end,
})
```

I applied the config suggested in this [Neovim issue](https://github.com/neovim/neovim/issues/27740) and tested that it works on 0.9.5 and nightly.

### CODE REVIEW

**Comment by abeldekat** [comment]
@MariaSolOs,

In this [PR](https://github.com/neovim/nvim-lspconfig/pull/3029) you added:

-- Depending on the usage, you might want to add additional paths here.
-- E.g.: For using `vim.*` functions, add vim.env.VIMRUNTIME/lua.

I tested with and without that addition and could not detect a difference.
Is that line still needed?

File: lua/lspconfig/server_configurations/lua_ls.lua [hunk 1]
```diff
@@ -47,31 +47,32 @@ require'lspconfig'.lua_ls.setup {
...
-            library = {
-              vim.env.VIMRUNTIME
-              -- Depending on the usage, you might want to add additional paths here.
-              -- E.g.: For using `vim.*` functions, add vim.env.VIMRUNTIME/lua.
-              -- "${3rd}/luv/library"
-              -- "${3rd}/busted/library",
-            }
-            -- or pull in all of 'runtimepath'. NOTE: this is a lot slower
-            -- library = vim.api.nvim_get_runtime_file("", true)
+      client.config.settings.Lua = vim.tbl_deep_extend('force', client.config.settings.Lua, {
```
Comment (lewis6991) [inline lua/lspconfig/server_configurations/lua_ls.lua:50]:
```suggestion
      client.settings.Lua = vim.tbl_deep_extend('force', client.settings.Lua, {
```

**Comment by lewis6991** [comment]
Yes, this change makes sense since `vim.lsp.Client` now has its own reference to `settings`.

File: lua/lspconfig/server_configurations/lua_ls.lua [hunk 2]
```diff
@@ -47,31 +47,32 @@ require'lspconfig'.lua_ls.setup {
...
-            library = {
-              vim.env.VIMRUNTIME
-              -- Depending on the usage, you might want to add additional paths here.
-              -- E.g.: For using `vim.*` functions, add vim.env.VIMRUNTIME/lua.
-              -- "${3rd}/luv/library"
-              -- "${3rd}/busted/library",
-            }
-            -- or pull in all of 'runtimepath'. NOTE: this is a lot slower
-            -- library = vim.api.nvim_get_runtime_file("", true)
+      client.settings.Lua = vim.tbl_deep_extend('force', client.settings.Lua, {
```
Comment (lewis6991) [inline lua/lspconfig/server_configurations/lua_ls.lua:50]: Just to note that `client.settings = vim.tbl_deep_extend(...)` will also work.

**Review by lewis6991 — APPROVED** [review]
Squash and fix the commit messages and we can merge.

**Comment by abeldekat** [comment]
@lewis6991,

I find the following hard to read:
```lua
    if not vim.loop.fs_stat(path..'/.luarc.json') and not vim.loop.fs_stat(path..'/.luarc.jsonc') then
      client.config.settings.Lua = vim.tbl_deep_extend('force', client.config.settings.Lua, {
```

As a suggestion:
```lua
require'lspconfig'.lua_ls.setup {
  on_init = function(client)
    local path = client.workspace_folders[1].name
    if vim.loop.fs_stat(path..'/.luarc.json') or vim.loop.fs_stat(path..'/.luarc.jsonc') then
      return true
    end

    client.settings.Lua = vim.tbl_deep_extend('force', client.settings.Lua, {
      runtime = {
        -- Tell the language server which version of Lua you're using
        -- (most likely LuaJIT in the case of Neovim)
        version = 'LuaJIT'
      },
      -- Make the server aware of Neovim runtime files
      workspace = {
        checkThirdParty = false,
        library = {
          vim.env.VIMRUNTIME
          -- Depending on the usage, you might want to add additional paths here.
          -- E.g.: For using `vim.*` functions, add vim.env.VIMRUNTIME/lua.
          -- "${3rd}/luv/library"
          -- "${3rd}/busted/library",
        }
        -- or pull in all of 'runtimepath'. NOTE: this is a lot slower
        -- library = vim.api.nvim_get_runtime_file("", true)
      }
    })
    end
    return true
  end
  settings = {
    Lua = {}
  }
}
```

Also: Is return true needed?

**Comment by lewis6991** [comment]
No it's not.

**Comment by abeldekat** [comment]
Then I will also remove the return true if you agree

What is your opinion on my early return suggestion?
```lua
if vim.loop.fs_stat(path..'/.luarc.json') or vim.loop.fs_stat(path..'/.luarc.jsonc') then
      return true
    end
```

**Comment by lewis6991** [comment]
Yeah, looks fine. Again don't need to return `true`.

**Comment by abeldekat** [comment]
@lewis6991, I added a commit.
Just one final question:

-- E.g.: For using vim.* functions, add vim.env.VIMRUNTIME/lua.

Is this still needed? I did not observe any difference.

**Comment by lewis6991** [comment]
No I don't think that does anything.

**Comment by abeldekat** [comment]
There is one problem: By using client.settings, the suggested config will not work on version `0.9.5` anymore...

**Comment by lewis6991** [comment]
Ok, don't use that for now.

**Comment by abeldekat** [comment]
I restored client.config.settings and removed the comment mentioning VIMRUNTIME/lua

> Squash and fix the commit messages and we can merge.

Do I need to do this or is this done when merging?

**Comment by abeldekat** [comment]
Done. Thanks, @lewis6991!

**Comment by MariaSolOs** [comment]
> @MariaSolOs,
> 
> In this [PR](https://github.com/neovim/nvim-lspconfig/pull/3029) you added:
> 
> -- Depending on the usage, you might want to add additional paths here. -- E.g.: For using `vim.*` functions, add vim.env.VIMRUNTIME/lua.
> 
> I tested with and without that addition and could not detect a difference. Is that line still needed?

Sorry for the late response. @lewis6991 is correct.

And thanks for the change!

**Pull request merged**